### ISSUE TITLE
Option which is optionally a flag (i.e optional value)

### ISSUE THREAD

**Issue by amirabiri** [ajalt/clikt#346]
I was looking to do something like this:

```
./tool --csv # Valid - output in CSV format to stdout
./tool --csv=filename.csv # Also valid - output in CSV format to the specified filename
```

However there doesn't seem to be a way of saying that an option can take 0 or 1 values. I also tried a group that includes `option("--csv").flag()...` with `option("--csv").path()...` but that didn't work.

**Comment by ajalt** [comment]
Options don't support variable numbers of arguments in the current version of Clikt. See #280 for a previous discussion.

You can try out the [4.0](https://github.com/ajalt/clikt/tree/4.0) branch, which has support for vararg options. With 4.0 you could implement your use case by passing a range to `transformValues` like this:

```kotlin
sealed class CsvValue {
    data class Path(val path: Path): CsvValue()
    object NoPath: CsvValue()
}
val csv: CsvValue by option().path().transformValues(0..1) {
    it?.let { CsvValue.Path(it) } ?: CsvValue.NoPath
}
```

Try that out and let me know if it works for you.

**Comment by amirabiri** [comment]
Version 4.0 has maven builds or not yet?

**Comment by ajalt** [comment]
Not yet, you'll have to build yourself.

**Comment by ajalt** [comment]
@amirabiri I set up [snapshot publishing for the 4.0 branch]. There are instructions in the readme one how to use them. The latest 4.0 snapshot version is `4.0.0.184-SNAPSHOT`. Let me know what you think.

[1]: https://oss.sonatype.org/content/repositories/snapshots/com/github/ajalt/clikt/clikt/4.0.0.184-SNAPSHOT/

**Comment by amirabiri** [comment]
Tested it and it works well.

The help generation hasn't caught up yet:

```
# ./command --help
Usage: command [OPTIONS]

Options:
  --csv PATH
  -h, --help  Show this message and exit
```

I would expect it to indicate that PATH is optional. But all the actual scenarios work well:

```
./command
./command --csv
./command --csv=file.txt
```

Plus there seems to be a problem with subcommands:

```
# ./command subcommand --help
Usage: command [OPTIONS] COMMAND [ARGS]...

Options:
  -h, --help  Show this message and exit

Commands:
  subcommand
```

I assume this is known?

Anyway none of it seems related to the core issue, which seems resolved by this to me.

**Comment by ajalt** [comment]
I fixed the subcommand help in the latest 4.0 snapshot, and showing the optional value in help output is on my todo list, so I'll close this issue as solved. Thanks for all the feedback!
